src/node.cc

```diff
@@ -303,6 +304,27 @@ MaybeLocal<Value> StartExecution(Environment* env, StartExecutionCallback cb) {
     first_argv = env->argv()[1];
   }
 
+#ifndef DISABLE_SINGLE_EXECUTABLE_APPLICATION
+  size_t single_executable_application_size = 0;
+  const char* single_executable_application_code =
+      FindSingleExecutableCode(&single_executable_application_size);
+  if (single_executable_application_code != nullptr) {
+    Local<Value> buffer =
+        Buffer::New(
+            env->isolate(),
+            const_cast<char*>(single_executable_application_code),
+            single_executable_application_size,
+            [](char* data, void* hint) {},
+            nullptr)
+            .ToLocalChecked();
+    env->process_object()
+        ->SetPrivate(
+            env->context(), env->single_executable_application_code(), buffer)
+        .Check();
+    return StartExecution(env, "internal/main/single_executable_application");
+  }
+#endif
```

Comment:
This is really just mostly re-inventing the wheel that `LoadEnvironment(Environment*, const char*)` already implements. If we can avoid it, I’d re-use that and not add yet another main entry point here (which, as the comments note, already duplicates existing code)

Reply:
@addaleax looking at https://github.com/nodejs/node/blob/3abd559c6a2a32c7d313ed1149f40d53fa77cc1f/doc/api/embedding.md?plain=1#L142-L143, it seems like the `require()` object will not able to access the filesystem, so it would probably still require something like what I have in `lib/internal/main/single_executable_application.js`. Wouldn't it be better to keep that in a file instead of putting it into a string in this file?

Reply:
@addaleax I just tried replacing the body of this if block with `return LoadEnvironment(env, single_executable_application_code);` and it looks like the `require()` that's exposed here is the internal require and that would mean that user code will be able to do things like `require('internal/...');`. So it means that we would still have to expose a different require to the code that gets run, right?

> require relative from the current executable

To be clear, I did that so that it is possible to require our 'test/common' module in the SEA so that I can easily verify the exposed globals, etc. :P I'd be happy to use a different approach if you could share one!

Reply:
> So it means that we would still have to expose a different require to the code that gets run, right?

Yes, and again, ideally that is also true for embedder cases – we shouldn’t really let them require() internal modules :)

> I'd be happy to use a different approach if you could share one!

Would it seem reasonable to you to let the user create a new require function via `createRequire()` if they want to access on-disk modules? That way they can directly specify which path the requires should be relative from, which we can also do fairly easily from our own tests

Reply:
> via `createRequire()`

Do you mean by exposing `createRequire` as a global or a parameter to the injected script? While that would certainly work, I think that would make it harder for most users to use this feature because of 2 reasons:
* `createRequire` isn't present globally or as a parameter by default on a regular Node.js script, so this would be unexpected to users.
* I don't think most users are accustomed with this function.

If users feel strongly about using a `require()` that uses a different base path, I guess they could use `createRequire` to create a new require and use that instead, which is possible in the current implementation. Wdyt?

Reply:
> Do you mean by exposing `createRequire` as a global or a parameter to the injected script?

No, for exactly the reasons you mentioned :)

And idk, I would still be very careful about this. Exposing an fs-based `require()` probably isn’t going to be what 99% of SEA users want, since one of the purposes of having an SEA is having all existing code bundled, so starting to load modules from the file system is something that should be done intentional.

(Imagine the following scenario: There’s a SEA X that contains an npm package that has an optional dependency that it loads via `try { require("asdf") } catch {}`. Somebody downloads X, and, because they are careless, they put it into `/tmp/somefolder` and run it from there. Somebody else who has access to the same system is a smart hacker and creates `/tmp/node_modules/asdf/index.js` with malicious code – the SEA X will just happily load that file in that case.)

Reply:
Hmm, one solution I can think of is to expose a `require()` that looks like this:
```js
function seaRequire(mod) {
  if (!Module.isBuiltin(mod)) {
    throw new ERR_UNKNOWN_BUILTIN_MODULE(mod);
  }
  return require(mod);
}
```
@addaleax what do you think of that approach?

Reply:
@RaisinTen Yeah, I think that would make sense both for embedders and for SEAs :+1:

Reply:
@addaleax I have pushed a commit to implement this. Does this address your concern correctly?

Reply:
I think it’s okay, yeah. It doesn’t really address the original concern here of building two different code paths to achieve the same result, but that can be something that can be addressed after the initial PR.